packages/sourcecred/src/api/main/grain.js

```diff
@@ -37,3 +41,30 @@ export async function grain(input: GrainInput): Promise<GrainOutput> {
   );
   return {distributions, ledger: input.ledger};
 }
+
+/**
+ * Marshall grainInput from a Grain Configuration file for use with
+ * executeGrainIntegration function
+ */
+export function executeGrainIntegrationsFromGrainInput(
+  grainInput: GrainInput,
+  ledger: Ledger,
+  distributions: $ReadOnlyArray<Distribution>
+): $ReadOnlyArray<GrainIntegrationResult> {
+  const integrationCurrency = grainInput.currencyDetails.integrationCurrency;
+  const grainIntegration = grainInput.grainConfig.integration;
+  const results = [];
+  if (integrationCurrency && grainIntegration) {
+    distributions.forEach((distribution) => {
+      const result = executeGrainIntegration(
+        ledger,
+        grainIntegration.function,
+        distribution,
+        integrationCurrency,
+        false
+      );
+      results.push(result);
+    });
+  }
+  return results;
+}
```

Comment:
Seems strange for this to return an array of objects that each contain the same ledger, and that in the CLI we encode the assumption that the ledger we passed in is going to be mutated and so we can ignore the ledgers that come out. What about a refactor that looks more like:
```suggestion
  if (integrationCurrency && grainIntegration) {
  let ledgerResult = ledger;
    distributions.forEach((distribution) => {
      const {ledger: newLedger, output, distributionCredTimestamp} = executeGrainIntegration(
        ledger,
        grainIntegration.function,
        distribution,
        integrationCurrency,
        false
      );
      results.push({output, distributionCredTimestamp});
      ledgerResult = newLedger;
    });
  }
  return {ledger: ledgerResult, results};
}
```

Reply:
yeah good catch. We don't ever want to assume it's safe to mutate the same ledger reference.